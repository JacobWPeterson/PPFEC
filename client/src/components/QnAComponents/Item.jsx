import React from 'react';
import styled from 'styled-components';
import AnswerList from './AnswerList.jsx';
import MoreA from './MoreA.jsx';
import Modal from './Modal.jsx';
import FormA from './FormAnswer.jsx';

const TopWrapper = styled.div`
  display: flex;
  flex-direction: row;
  font-size: 20px;
  padding: 2vh 0;
  justify-content: space-between;
  align-items: center;
`;

const BottomWrapper = styled.div`
  display: flex;
  flex-direction: row;
`;

const Helper = styled.div`
  display: flex;
  font-size: 15px;
  min-width: 245px;
  padding-left: 4vw;
  color: #424242;
`;

const Yes = styled.div`
  padding: 0 2px 0 1vw;
  text-decoration: underline;
  &:hover {
    cursor: pointer;
    color: #80CCC4;
    transform: scale(1.1);
  };
`;

const AddAButton = styled.div`
  padding-left: 1vw;
  text-decoration: underline;
  &:hover {
    cursor: pointer;
    color: #80CCC4;
    transform: scale(1.1);
  };
`;

const Helpfulness = styled.div`
  padding-right: 1vw;
`;

const A = styled.div`
  font-size: 20px;
  padding-right: 1vw;
  font-weight: 700;
  color: #424242;
`;

const Q = styled.div`
  font-weight: 700;
  padding-right: 1vw;
  color: #424242;
`;

const QB = styled.h4`
  display: flex;
  flex-direction: row;
  font-weight: 700;
  font-size: 1.3rem;
`;

const ListContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

class Item extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      answers: [],
      dd: [],
      max: 0,
      displayed: 0,
      buttonDisplay: true,
      modal: false,
    };
    this.loadTwoItems = this.loadTwoItems.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.exitModal = this.exitModal.bind(this);
  }

  componentDidMount() {
    Object.entries(this.props.item.answers).forEach((answer) => {
      this.state.answers.push(answer[1]);
    });
    this.setState((prevState) => ({
      answers: prevState.answers,
    }));
    this.loadTwoItems();
  }

  handleClick() {
    this.setState({ modal: true });
  }

  exitModal() {
    this.setState({ modal: false });
  }

  loadTwoItems() {
    this.state.max += 2;
    while (this.state.displayed < this.state.max) {
      if (!this.state.answers[this.state.displayed]) {
        break;
      }
      this.state.dd.push(this.state.answers[this.state.displayed]);
      this.state.displayed += 1;
    }
    if (!this.state.answers[this.state.displayed]) {
      this.state.buttonDisplay = false;
    }
    this.setState((prevState) => ({
      dd: prevState.dd,
      displayed: prevState.displayed,
      max: prevState.max,
      buttonDisplay: prevState.buttonDisplay,
    }));
  }

  render() {
    return (
      <div>
        <TopWrapper>
          <QB>
            <Q>Q:</Q>
            {this.props.item.question_body}
          </QB>
          <Helper>
            <div>Helpful?</div>
            <Yes
              onClick={this.handleClick}
            >
              Yes
            </Yes>
            {this.state.modal ? (
              <Modal content={(
                <FormA
                  func={this.exitModal}
                />
              )}
              />
            ) : null}
            <Helpfulness>{`(${this.props.item.question_helpfulness})`}</Helpfulness>
            <div>|</div>
            <AddAButton
              onClick={this.handleClick}
            >
              Add Answer
            </AddAButton>
          </Helper>
        </TopWrapper>
        <BottomWrapper>
          <A>A:</A>
          <ListContainer>
            <AnswerList answers={this.state.dd} />
            <MoreA func={this.loadTwoItems} display={this.state.buttonDisplay} />
          </ListContainer>
        </BottomWrapper>
      </div>
    );
  }
}

export default Item;
