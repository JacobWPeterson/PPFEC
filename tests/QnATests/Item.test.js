import React from 'react';
import { shallow, mount } from 'enzyme';
import Item from '../../client/src/components/QnAComponents/Item.jsx';

describe('One Question Test', () => {
  const props = {
    id: 152944,
    item: {
      answers: {
        1443618: {
          answerer_name: 'sadf',
          body: 'asdf',
          date: '2021-03-08T00:00:00.000Z',
          helpfulness: 0,
          id: 1443618,
          photos: ['https://pngimg.com/uploads/mario/mario_PNG53.png', 'https://pngimg.com/uploads/mario/mario_PNG53.png'],
        },
        asker_name: 'asdf',
        question_body: 'awdsf',
        question_date: '2021-03-08T00:00:00.000Z',
        question_helpfulness: 1,
        question_id: 152944,
        reported: false,
      },
    },
  };

  const itemComp = shallow(<Item {...props} debug />);
  const itemBody = (itemComp.find('div').at(0).text());

  it('Should render Item correctly', () => {
    expect(itemComp).toMatchSnapshot();
  });

  it('calls handleClick', () => {
    expect(itemBody).toEqual('Q:Helpful?Yes(undefined)|Add AnswerA:<AnswerList /><MoreA />');
  });

  // it('Should render the answer body correctly', () => {
  //   expect(itemHelp).toEqual('Helpful?');
  // });

  // it('Should render the answer body correctly', () => {
  //   expect(itemSep).toEqual('|');
  // });
});
