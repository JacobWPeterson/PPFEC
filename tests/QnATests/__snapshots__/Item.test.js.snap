// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`One Question Test Should render Item correctly 1`] = `
<<<<<<< HEAD
<div>
  <Item__TopWrapper>
    <Item__QB>
      <Item__Q>
        Q:
      </Item__Q>
    </Item__QB>
    <Item__Helper>
      <div>
        Helpful?
      </div>
      <Item__Yes
        onClick={[Function]}
      >
        Yes
      </Item__Yes>
      <Item__Helpfulness>
        (undefined)
      </Item__Helpfulness>
      <div>
        |
      </div>
      <Item__AddAButton
        onClick={[Function]}
      >
        Add Answer
      </Item__AddAButton>
    </Item__Helper>
  </Item__TopWrapper>
  <Item__BottomWrapper>
    <Item__A>
      A:
    </Item__A>
    <Item__ListContainer>
      <AnswerList
        answers={
          Array [
            Object {
              "answerer_name": "sadf",
              "body": "asdf",
              "date": "2021-03-08T00:00:00.000Z",
              "helpfulness": 0,
              "id": 1443618,
              "photos": Array [
                "https://pngimg.com/uploads/mario/mario_PNG53.png",
                "https://pngimg.com/uploads/mario/mario_PNG53.png",
              ],
            },
            "asdf",
          ]
        }
      />
      <MoreA
        display={true}
        func={[Function]}
      />
    </Item__ListContainer>
  </Item__BottomWrapper>
</div>
`;

exports[`One Question Test Should render correctly 1`] = `
=======
>>>>>>> 02ee0ec6c214bf9bfe9c4a4496ac203921d04516
<div>
  <Item__TopWrapper>
    <Item__QB>
      <Item__Q>
        Q:
      </Item__Q>
    </Item__QB>
    <Item__Helper>
      <div>
        Helpful?
      </div>
      <Item__Yes
        onClick={[Function]}
      >
        Yes
      </Item__Yes>
      <Item__Helpfulness>
        (undefined)
      </Item__Helpfulness>
      <div>
        |
      </div>
      <Item__AddAButton
        onClick={[Function]}
      >
        Add Answer
      </Item__AddAButton>
    </Item__Helper>
  </Item__TopWrapper>
  <Item__BottomWrapper>
    <Item__A>
      A:
    </Item__A>
    <Item__ListContainer>
      <AnswerList
        answers={
          Array [
            Object {
              "answerer_name": "sadf",
              "body": "asdf",
              "date": "2021-03-08T00:00:00.000Z",
              "helpfulness": 0,
              "id": 1443618,
              "photos": Array [
                "https://pngimg.com/uploads/mario/mario_PNG53.png",
                "https://pngimg.com/uploads/mario/mario_PNG53.png",
              ],
            },
            "asdf",
          ]
        }
      />
      <MoreA
        display={true}
        func={[Function]}
      />
    </Item__ListContainer>
  </Item__BottomWrapper>
</div>
`;
